types/opt/value.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package opt
+
+import (
+	"fmt"
+	"reflect"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+)
+
+// Value is an optional value to be JSON-encoded.
+// Unlike the specialized Bool type, an empty (unset) Value[T] is marshaled as
+// a JSON null regardless of the "omitempty" JSON tag option when using [encoding/json].
+// With [github.com/go-json-experiment/json], however, empty optional Values are
+// omitted from the output.
+type Value[T any] struct {
+	value T
+	set   bool
+}
+
+// Equal reports whether the receiver and the other value are equal.
+// If the template type T in Value[T] implements an Equal method, it will be used
+// instead of the == operator for comparing values.
+type equatable[T any] interface {
+	// Equal reports whether the receiver and the other values are equal.
+	Equal(other T) bool
+}
+
+// Empty returns an empty optional Value.
+func Empty[T any]() Value[T] {
+	return Value[T]{set: false}
+}
+
+// Of returns an optional Value containing the specified value.
+// It treats nil pointers, unsafe pointers, and interfaces as unset values,
+// but nil slices and maps as empty slices and maps.
+func Of[T any](v T) Value[T] {
+	switch rv := reflect.ValueOf(v); rv.Kind() {
```

Comment:
Removed for maps, slices, and interfaces. However, I would prefer to keep this behavior for pointers unless there are any objections. I think it might simplify usage at call sites.

Reply:
I think my main objection comes from the fact that `Of` doesn't always return a value that is populated.

`Of([]T(nil))` and  `Of((*T)(nil))` differ in whether the returned `Value` where `Value.Has` differs.

Reply:
Agreed. Let's keep it consistent.